pkg/nb/rpc.go

```diff
@@ -122,19 +205,251 @@ func (c *RPCClient) Call(req *RPCRequest, res RPCResponseIfc) error {
+func (r *RPC) GetConnection(pending *RPCPendingRequest, address string) error {
+	r.ConnMapLock.Lock()
+	pending.Conn = r.ConnMap[address]
+	if pending.Conn == nil {
+		ws, _, err := websocket.Dial(context.TODO(), address, &websocket.DialOptions{HTTPClient: &r.HTTPClient})
+		if err != nil {
+			r.ConnMapLock.Unlock()
+			return err
+		}
+		pending.Conn = &RPCConn{
+			WS:              ws,
+			PendingRequests: map[string]*RPCPendingRequest{},
+		}
+		r.ConnMap[address] = pending.Conn
+		go r.HandleConnection(pending.Conn, address)
+	}
+	pending.Req.RequestID = fmt.Sprintf("%s-%d", address, pending.Conn.NextRequestID)
+	pending.Conn.NextRequestID++
+	pending.Conn.PendingRequests[pending.Req.RequestID] = pending
+	r.ConnMapLock.Unlock()
+	return nil
+}
+
+// SendMessage sends the pending request
+func (r *RPC) SendMessage(conn *RPCConn, msg interface{}) error {
+	writer, err := conn.WS.Writer(context.TODO(), websocket.MessageBinary)
 	if err != nil {
-		panic(err)
+		return err
+	}
+
+	reqBytes, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	err = binary.Write(writer, binary.BigEndian, RPCVersionNumber)
+	if err != nil {
+		return err
+	}
+
+	err = binary.Write(writer, binary.BigEndian, uint32(len(reqBytes)))
+	if err != nil {
+		return err
+	}
+
+	_, err = writer.Write(reqBytes)
+	if err != nil {
+		return err
+	}
+
+	err = writer.Close()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ConnectWs tries to connect to address on web socket connection
+func (r *RPC) ConnectWs(address string) (*websocket.Conn, error) {
+	ws, _, err := websocket.Dial(context.TODO(), address, &websocket.DialOptions{HTTPClient: &r.HTTPClient})
+	if err != nil {
+		return nil, fmt.Errorf("RPC: connection failed")
+	}
+	return ws, nil
+}
+
+// WaitResponse waits for the response of the pending request
+func (r *RPC) WaitResponse(pending *RPCPendingRequest) error {
+	return <-pending.CloseChan
+}
+
+// ReadMessage handles a message
+func (r *RPC) ReadMessage(conn *RPCConn, address string) (*RPCMessage, error) {
+	_, reader, err := conn.WS.Reader(context.TODO())
+	if err != nil {
+		if websocket.CloseStatus(err) == websocket.StatusNormalClosure {
+			_, err1 := r.ConnectWs(address)
+			if err1 != nil {
+				return nil, fmt.Errorf("RPC: connection closed, reconnecting failed")
+			}
+			return nil, fmt.Errorf("RPC: connection closed, start reconnecting")
+			// break
+			// reconnect
+		}
+		util.Panic(err)
+	}
+	var rpcVersionNumber uint32
+	var bodySize uint32
+
+	if err := binary.Read(reader, binary.BigEndian, &rpcVersionNumber); err != nil {
+		return nil, err
+	}
+	if err := binary.Read(reader, binary.BigEndian, &bodySize); err != nil {
+		return nil, err
+	}
+	msgBytes := make([]byte, bodySize)
+	n, err := reader.Read(msgBytes)
+	if err != nil {
+		if websocket.CloseStatus(err) == websocket.StatusNormalClosure {
+			fmt.Printf("websocket closed, err: %+v\n", err)
+			_, err1 := r.ConnectWs(address)
```

Comment:
We probably want a short delay before reconnect to avoid tight loops when the address is wrong or there is no network available...